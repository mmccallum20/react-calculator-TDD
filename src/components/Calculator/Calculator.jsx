import React, { Component } from "react";
import Display from "../Display/Display";
import "./Calculator.css";
import KeyPad from "../Keypad/Keypad";

class Calculator extends Component {
  state = {
    //value to be displayed in <Display />
    displayValue: "0",

    //values to be displayed in number <Keys />
    numbers: ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", ".", "CE"],

    //values to be displayed in operator <Keys />
    operators: ["+", "-", "*", "/"],

    //operator selected for math operation
    selectedOperator: "",

    //stored value to use for math operation
    storedValue: "",
  };

  callOperator = () => {
    console.log("call operation");
  };

  setOperator = () => {
    console.log("set operation");
  };

  updateDisplay = () => {
    console.log("update display");
  };

  render = () => {
    const { displayValue, numbers, operators } = this.state;

    return (
      <div className="calculator-container">
        <Display displayValue={displayValue} />
        <KeyPad
          callOperator={this.callOperator}
          numbers={numbers}
          operators={operators}
          setOperator={this.setOperator}
          updateDisplay={this.updateDisplay}
        />
      </div>
    );
  };
}

export default Calculator;
