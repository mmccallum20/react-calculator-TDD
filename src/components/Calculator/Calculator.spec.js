import React from "react";
import { mount } from "enzyme";
import Calculator from "./Calculator";
import Display from "../Display/Display";
import KeyPad from "../Keypad/Keypad";

describe("mounted Calculator", () => {
  let wrapper;

  //shallow refers to only the component being tested
  //mount refers to the component and all its child components

  beforeEach(() => (wrapper = mount(<Calculator />)));

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls updateDisplay when a number key is clicked", () => {
    //spyOn is a feature that allows dynamically intercepting the calls to a function and changes its result
    const spy = jest.spyOn(wrapper.instance(), "updateDisplay");

    //forceUpdate triggers the normal lifecycle methods for child components
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);

    //click is simulated
    wrapper.find(".number-key").first().simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("calls setOperator when an operator key is clicked", () => {
    //spyOn is a feature that allows dynamically intercepting the calls to a function and changes its result
    const spy = jest.spyOn(wrapper.instance(), "setOperator");

    //forceUpdate triggers the normal lifecycle methods for child components
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);

    //click is simulated
    wrapper.find(".operator-key").first().simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("calls callOperator when a submit key is clicked", () => {
    //spyOn is a feature that allows dynamically intercepting the calls to a function and changes its result
    const spy = jest.spyOn(wrapper.instance(), "callOperator");

    //forceUpdate triggers the normal lifecycle methods for child components
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);

    //click is simulated
    wrapper.find(".submit-key").first().simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should render the Display and Keypad components", () => {
    expect(
      wrapper.containsAllMatchingElements([
        <Display displayValue={wrapper.instance().state.displayValue} />,
        <KeyPad
          callOperator={wrapper.instance().callOperator}
          numbers={wrapper.instance().state.numbers}
          operators={wrapper.instance().state.operators}
          setOperator={wrapper.instance().setOperator}
          updateDisplay={wrapper.instance().updateDisplay}
        />,
      ])
    ).toEqual(true);
  });

  it("should render a <div />", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });
});
