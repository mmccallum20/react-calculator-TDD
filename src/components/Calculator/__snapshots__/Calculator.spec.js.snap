// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mounted Calculator should render correctly 1`] = `
<Calculator>
  <div
    className="calculator-container"
  >
    <Display
      displayValue="0"
    >
      <div
        className="display-container"
      >
        <p
          className="display-value"
        >
          0
        </p>
      </div>
    </Display>
    <KeyPad
      callOperator={[Function]}
      numbers={
        Array [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          ".",
          "CE",
        ]
      }
      operators={
        Array [
          "+",
          "-",
          "*",
          "/",
        ]
      }
      setOperator={[Function]}
      updateDisplay={[Function]}
    >
      <div
        className="keypad-container"
      >
        <div
          className="numbers-container"
        >
          <Key
            key="0"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="0"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                0
              </p>
            </div>
          </Key>
          <Key
            key="1"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="1"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                1
              </p>
            </div>
          </Key>
          <Key
            key="2"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="2"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                2
              </p>
            </div>
          </Key>
          <Key
            key="3"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="3"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                3
              </p>
            </div>
          </Key>
          <Key
            key="4"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="4"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                4
              </p>
            </div>
          </Key>
          <Key
            key="5"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="5"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                5
              </p>
            </div>
          </Key>
          <Key
            key="6"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="6"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                6
              </p>
            </div>
          </Key>
          <Key
            key="7"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="7"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                7
              </p>
            </div>
          </Key>
          <Key
            key="8"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="8"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                8
              </p>
            </div>
          </Key>
          <Key
            key="9"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="9"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                9
              </p>
            </div>
          </Key>
          <Key
            key="."
            keyAction={[Function]}
            keyType="number-key"
            keyValue="."
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                .
              </p>
            </div>
          </Key>
          <Key
            key="CE"
            keyAction={[Function]}
            keyType="number-key"
            keyValue="CE"
          >
            <div
              className="key-container number-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                CE
              </p>
            </div>
          </Key>
        </div>
        <div
          className="operators-container"
        >
          <Key
            key="+"
            keyAction={[Function]}
            keyType="operator-key"
            keyValue="+"
          >
            <div
              className="key-container operator-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                +
              </p>
            </div>
          </Key>
          <Key
            key="-"
            keyAction={[Function]}
            keyType="operator-key"
            keyValue="-"
          >
            <div
              className="key-container operator-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                -
              </p>
            </div>
          </Key>
          <Key
            key="*"
            keyAction={[Function]}
            keyType="operator-key"
            keyValue="*"
          >
            <div
              className="key-container operator-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                *
              </p>
            </div>
          </Key>
          <Key
            key="/"
            keyAction={[Function]}
            keyType="operator-key"
            keyValue="/"
          >
            <div
              className="key-container operator-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                /
              </p>
            </div>
          </Key>
        </div>
        <div
          className="submit-container"
        >
          <Key
            keyAction={[Function]}
            keyType="submit-key"
            keyValue="="
          >
            <div
              className="key-container submit-key"
              onClick={[Function]}
            >
              <p
                className="key-value"
              >
                =
              </p>
            </div>
          </Key>
        </div>
      </div>
    </KeyPad>
  </div>
</Calculator>
`;
